### PR TITLE
proxy_proto: fixing hashing bug

### LINKED ISSUE #13659: Proxy protocol transport socket intermittently injects wrong addresses
*Title*: *Proxy protocol header intermittently uses wrong addresses*

*Description*:
Proxy protocol injection (envoy.transport_sockets.upstream_proxy_protocol) works as expected to start out, but once you start connecting from different IPs, the injected proxy protocol header ends up randomly having addresses from previous connections.
The log shows the right IPs but proxy protocol header intermittently sends the IPs not from the current connection, but from one of the previous ones.
I thought maybe setting `connection_pool_per_downstream_connection: true` would help but it did not.

*Repro steps*:
Have envoy listen on 127.0.0.1:9000 and tcp proxy to a cluster that listens for connections, reads the proxy protocol header, and then promptly closes the connection.
```
# run this about 20 times:
curl --interface 127.0.0.2 127.0.0.1:9000
# then run this about 20 times
curl --interface 127.0.0.3 127.0.0.1:9000
# then run this about 20 times
curl --interface 127.0.0.4 127.0.0.1:9000
```
The expectation here and what the logs show is 20 connections each from the 3 separate addresses above. What the proxy protocol sends, ends up being a random mix.

*Suspected cause*:
I'm thinking this has something to do with connection pooling as the following comment suggests that it's not exactly a 1 to 1 setup currently: https://github.com/envoyproxy/envoy/pull/4684#discussion_r224930480

*Potentially related*: https://github.com/envoyproxy/envoy/issues/2800 https://github.com/envoyproxy/envoy/issues/11694

### PR DESCRIPTION
**Pull request by alyssawilk**
Fix a bug where the transport socket options for the first downstream got reused for subsequent upstream connections.

Risk Level: low
Testing: new integration test
Docs Changes: n/a
Release Notes:
Platform Specific Features:
Fixes #13659

### CODE REVIEW

**Comment by alyssawilk** [comment]
Also per @rchernobelskiy 's comment on the issue, can anyone think of a better way to avoid bugs of this type?  I really dislike sizeof tests, but the best I can think of is a sizeof() test which best-effort catches new proxy protocol additions and suggests that we update the hash before adjusting the size.  Any better ideas?

File: test/integration/integration_tcp_client.h [hunk 1]
```diff
@@ -28,7 +28,9 @@ class IntegrationTcpClient {
 public:
   IntegrationTcpClient(Event::Dispatcher& dispatcher, MockBufferFactory& factory, uint32_t port,
                        Network::Address::IpVersion version, bool enable_half_close,
-                       const Network::ConnectionSocket::OptionsSharedPtr& options);
+                       const Network::ConnectionSocket::OptionsSharedPtr& options,
+                       Network::Address::InstanceConstSharedPtr source_address =
+                           Network::Address::InstanceConstSharedPtr());
```
Comment (lizan) [inline test/integration/integration_tcp_client.h:33]: style nit: this is nullptr?

**Review by lizan — COMMENTED** [review]
LGTM

File: include/envoy/network/transport_socket.h [hunk 2]
```diff
@@ -226,6 +229,11 @@ class TransportSocketFactory {
    */
   virtual TransportSocketPtr
   createTransportSocket(TransportSocketOptionsSharedPtr options) const PURE;
+
+  /**
+   * @return bool whether the transport socket will use proxy protocol options.
+   */
+  virtual bool usesProxyProtocolOptions() const { return false; }
```
Comment (lizan) [inline include/envoy/network/transport_socket.h:236]: tap transport socket would need return underlying transport socket value instead of false.

**Comment by alyssawilk** [comment]
Sorry, I think I need another stamp after merge

**Review by lizan — APPROVED** [review]

**Review by lizan — APPROVED** [review]

**Comment by rchernobelskiy** [comment]
Thanks you guys for the fix 🎉 
Do you know if it's planned for 1.16.1 soon or not for a while?

**Comment by alyssawilk** [comment]
@envoyproxy/stable-maintainers haven't reviewed the backport yet, but there's also not a set schedule for backport releases other than when we cut a CVE release.
@envoyproxy/maintainers were just discussing cadence here but currently I'm not sure we have enough folks signed on for stable release work to justify more than cutting a release every few months.

**Comment by cpakulski** [comment]
I think it should be backported to 1.16 - starting this work.

**Comment by cpakulski** [comment]
Backported to 1.16. Removing `backport/approved` label.

**Pull request merged**